### PR TITLE
NEMU 4.0 rc0 Rebase

### PR DESCRIPTION
**Pull request by yangzhon**
@sameo @rbradford @sboeuf 

Please help me review these 166 patches, there are lots of changes between Qemu 4.0 and Qemu 3.1. Please focus on pvh, loader, acpi(RSDP) and flash related files, thanks!

This PR has been verified in virt, pc and q35 platforms and all can bootup.

#nemu porting patch history
##. removed patch list
1)0007-target-arm-Move-all-VFP-helpers-into-their-own-file.patch
old commit id: 49b4228
reason: there is one vfp_helper.c in target/arm of QEMU4.0
(2)0046-hw-arm-Switch-to-the-AML-build-RSDP-building-routine.patch
old commit id: 1759f04
reason: build_rsdp() has been moved to aml-build.c in QEMU4.0

##. new patch added list
(1)0030: i386: virt refactor load_linux with conf struct
    commit id: 764c7e5
    reason: PVH support
(2) 0131: NVDIMM change
    commit is: 69239dc0d8648
    reason: the qemu 4.0 has moved nvdimm property from PC machine to MACHINE for all archs.

##heavy changed patches
(1) 0043: hw: fw-build: Add firmware build methods and state
(2) 0047: hw: acpi: Implement XSDT support for RSDP
(3) 0034: i386: virt: Load firmware on virt platform
     commit id: 1662d18443b4 
     I still kept hw/i386/pc_sysfw.c file because this file added lots of flash changes. I only changed this 
     old_pc_system_rom_init() name and virt will directly call this to load firmware. for PFLASH, i will create 
     one new file, which will only include  pc_system_rom_init() fucntion and virt platform will build out 
     PFLASH code.
     -static void old_pc_system_rom_init(MemoryRegion *rom_memory, bool isapc_ram_fw)
    +void pc_system_rom_init(MemoryRegion *rom_memory, bool isapc_ram_fw)

### CODE REVIEW

**Comment by rbradford** [comment]
@yangzhon ARM compile error:

```
  CC      x86_64-softmmu/hw/i386/virt/memory.o
In file included from /home/nemu-ci/jenkins-workspace/workspace/nemu_PR-213/include/qemu/osdep.h:51:0,
                 from /home/nemu-ci/jenkins-workspace/workspace/nemu_PR-213/hw/arm/virt-acpi-build.c:29:
/home/nemu-ci/jenkins-workspace/workspace/nemu_PR-213/include/qemu/compiler.h:74:35: error: invalid operands to binary - (have ‘_Bool *’ and ‘uint8_t * {aka unsigned char *}’)
 #define type_check(t1,t2) ((t1*)0 - (t2*)0)
                                   ^
/home/nemu-ci/jenkins-workspace/workspace/nemu_PR-213/include/migration/vmstate.h:235:6: note: in expansion of macro ‘type_check’
      type_check(_type, typeof_field(_state, _field)))
      ^
/home/nemu-ci/jenkins-workspace/workspace/nemu_PR-213/include/migration/vmstate.h:282:21: note: in expansion of macro ‘vmstate_offset_value’
     .offset       = vmstate_offset_value(_state, _field, _type),     \
                     ^
/home/nemu-ci/jenkins-workspace/workspace/nemu_PR-213/include/migration/vmstate.h:748:5: note: in expansion of macro ‘VMSTATE_SINGLE_TEST’
     VMSTATE_SINGLE_TEST(_field, _state, NULL, _version, _info, _type)
     ^
/home/nemu-ci/jenkins-workspace/workspace/nemu_PR-213/include/migration/vmstate.h:780:5: note: in expansion of macro ‘VMSTATE_SINGLE’
     VMSTATE_SINGLE(_f, _s, _v, vmstate_info_bool, bool)
     ^
/home/nemu-ci/jenkins-workspace/workspace/nemu_PR-213/include/migration/vmstate.h:801:5: note: in expansion of macro ‘VMSTATE_BOOL_V’
     VMSTATE_BOOL_V(_f, _s, 0)
     ^
/home/nemu-ci/jenkins-workspace/workspace/nemu_PR-213/hw/arm/virt-acpi-build.c:848:9: note: in expansion of macro ‘VMSTATE_BOOL’
         VMSTATE_BOOL(patched, AcpiBuildState),
         ^
  CC      aarch64-softmmu/qapi/qapi-types.o
/home/nemu-ci/jenkins-workspace/workspace/nemu_PR-213/rules.mak:69: recipe for target 'hw/arm/virt-acpi-build.o' failed
make[1]: *** [hw/arm/virt-acpi-build.o] Error 1

```

**Comment by yangzhon** [comment]
@rbradford

The compile error you mentioned in ARM has been fixed, but i found another ARM TCG related errors. Let me check this. thanks for help!

File: hw/pci-host/pci_lite.c [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+{
+    DeviceClass *dc = DEVICE_CLASS(klass);
+    PCIDeviceClass *k = PCI_DEVICE_CLASS(klass);
+
+    k->class_id = PCI_CLASS_BRIDGE_HOST;
+    dc->desc = "Host bridge";
+
+    // TODO: Use different one to GPEX?
+    k->vendor_id = PCI_VENDOR_ID_REDHAT;
+    k->device_id = PCI_DEVICE_ID_REDHAT_PCIE_HOST;
```
Comment (rbradford) [inline hw/pci-host/pci_lite.c]: I think a patch has gone missing to switch this over to our PCI ID.

Comment (yangzhon) [reply hw/pci-host/pci_lite.c]: This rebase only got to whole HW/ configurable for minimal CONFIGs(133b02b6), and the PCI ID changes(726b31) is after HW configurable patch series. So, don't worry, i will merge this patch after current ARM build issues are fixed.

**Comment by rbradford** [comment]
Superseded by #215 which rebases on 4.0rc4 and fixes the final remaining ARM build issues. Thanks for all the hard work on this @yangzhon

**Pull request closed without merging**